src/librbd/mirror/snapshot/PromoteRequest.cc

```diff
@@ -109,6 +109,11 @@ void PromoteRequest<I>::handle_list_watchers(int r) {
   }
 
   if (m_watchers.empty()) {
+    if(util::can_create_primary_snapshot(m_image_ctx, false, true,
```

Comment:
It seems like a little different approach here could be just to restart, i.e. call `send` method again so it would call `can_create_primary_snapshot`, do necessary checks and go without creating orphan this time.

Reply:
@trociny If `send` is called again, are you sure it wouldn't create a second orphan?  It seems that the namespace would still be in `MIRROR_SNAPSHOT_STATE_NON_PRIMARY` state which means that `can_create_primary_snapshot` would report that an orphan is needed.

Reply:
Then may be this could be improved? E.g. currently `*requires_orphan = !mirror_ns->is_demoted()` in `can_create_primary_snapshot` inside `if (mirror_ns->state == cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY)` block will always set it to `true`, so testing 
`!mirror_ns->is_demoted()` looks useless. Instead we could test if the current snapshot (`mirror_ns`) is orhan and do not set `requires_orphan` if it is orphan (i.e. we already created one). 

Note, this is just an idea and I did not evaluate it well, so I might be wrong. In general, my idea is that we should always be prepared to a request to be interrupted at any time (e.g. due to the application crash) and the user may try it again (restart) and it still works as expected. I am pessimistic that we handle correctly all the cases, but this might be a good opportunity to make this particular case to work as expected. But I am fine with any other approach to fix this.

Reply:
I agree.  The "works as expected after a crash" expectation is valid and we should be definitely be striving towards that goal.   Another case to consider here is two concurrent promote requests: the winning one should do the promote and succeed, the losing one should behave as if the image is already promoted.